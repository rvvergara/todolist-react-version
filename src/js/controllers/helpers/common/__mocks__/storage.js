const getDataFromLocalStorage = key => key;

export {
  getDataFromLocalStorage,
};
