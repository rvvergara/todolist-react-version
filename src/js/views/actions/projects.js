import uuid from 'uuid';

const getProjects = projects => ({
  type: 'GET_PROJECTS',
  projects,
});

const addProject = ({ name = 'PROJECT', description = 'SOME PROJECT' } = {}) => ({
  type: 'ADD_PROJECT',
  project: {
    id: uuid(),
    name,
    description,
    todos: [],
  },
});

const updateProject = (id, updates) => ({
  type: 'UPDATE_PROJECT',
  id,
  updates,
});

const deleteProject = id => ({
  type: 'DELETE_PROJECT',
  id,
});

export {
  getProjects,
  addProject,
  updateProject,
  deleteProject,
};
