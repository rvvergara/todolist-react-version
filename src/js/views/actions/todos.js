import uuid from 'uuid';

const addTodo = (todo, projectName) => ({
  type: 'ADD_TODO',
  todo: {
    ...todo,
    projectName,
    id: uuid(),
  },
});

const updateTodo = (id, updates) => ({
  type: 'UPDATE_TODO',
  id,
  updates,
});

const deleteTodo = id => ({
  type: 'DELETE_TODO',
  id,
});

export {
  addTodo,
  updateTodo,
  deleteTodo,
};
