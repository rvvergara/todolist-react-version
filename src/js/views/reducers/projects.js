export default (state = [], action) => {
  switch (action.type) {
    case 'GET_PROJECTS':
      return [...state, ...action.projects]

    case 'ADD_PROJECT':
    {
      const index = state.findIndex(project => project.name === action.project.name);
      return index === -1 ? state.concat(action.project) : state;
    }

    case 'UPDATE_PROJECT':
    {
      const { name } = action.updates;
      const conflictIndex = state.findIndex(project => project.name === name);
      const projectIndex = state.findIndex(project => project.id === action.id);
      const updatedProject = { ...state[projectIndex], ...action.updates };
      return conflictIndex === -1 ? [...state.slice(0, projectIndex), updatedProject, ...state.slice(projectIndex + 1)] : state;
    }
    case 'DELETE_PROJECT':
      return state.filter(project => project.id !== action.id);
    default:
      return state;
  }
};
