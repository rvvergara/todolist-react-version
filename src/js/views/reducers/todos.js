export default (state = [], action) => {
  switch (action.type) {
    case 'ADD_TODO':
      return [...state, action.todo];
    case 'UPDATE_TODO':
    {
      const todoIndex = state.findIndex(todo => todo.id === action.id);
      const updatedTodo = { ...state[todoIndex], ...action.updates };
      return [...state.slice(0, todoIndex), updatedTodo, ...state.slice(todoIndex + 1)];
    }
    case 'DELETE_TODO':
      return state.filter(todo => todo.id !== action.id);
    default:
      return state;
  }
};
