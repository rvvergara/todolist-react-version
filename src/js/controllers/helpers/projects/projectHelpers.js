import uuid from 'uuid';

import projectsController from '../../projectsController';

import * as localStorageData from '../common/storage';

export const createProjectsArray = () => {
  if (localStorageData.getDataFromLocalStorage('projectsArray') === null) {
    const projectsArray = [];
    localStorageData.setDataIntoLocalStorage('projectsArray', projectsArray);
  }
};

// Create a default project for new users or after localstorage is cleared
export const createDefaultProject = () => {
  if (localStorageData.getDataFromLocalStorage('projectCount') === 0) {
    const defaultProject = projectsController.create('Default Project', uuid());
  }
  if (localStorageData.getDataFromLocalStorage('Default Project') !== null) {
    const defaultProject = localStorageData.getDataFromLocalStorage('Default Project');
  }
};

// Adding event listener to Project Form
// document.getElementById('projectNameForm').addEventListener('submit', e => {
//   e.preventDefault();
//   submitProjectForm(e.target);
// });
// // Adding event listener to Add Project Btn
// document.getElementById('addProjBtn').addEventListener('click', e => {
//   const action = 'new';
//   showProjectForm(e.target, action);
// });
