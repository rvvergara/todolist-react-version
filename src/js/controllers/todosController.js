import Todo from '../models/todo';
import {
  updateProjectsArray,
  pushTodoToProject,
} from './helpers/todos/todoHelpers';

import * as localStorageData from './helpers/common/storage';

const todoController = (
  () => ({
    create(todoEntry) {
      const todo = new Todo(todoEntry);
      pushTodoToProject(todo);
      return todo;
    },

    update(id, updates) {
      const projectsArray = localStorageData.getDataFromLocalStorage('projectsArray');
      const todosArray = localStorageData.getDataFromLocalStorage('todosArray');
      const todoFromArray = todosArray.find(todo => todo.id === id);
      const projectInArray = projectsArray.find(project => project.id === todoFromArray.projectID);
      const projectIndexInArray = projectsArray.findIndex(project => project.id === projectInArray.id);
      const projectStandAlone = localStorageData.getDataFromLocalStorage(projectInArray.name);
      const todoIndexInProject = projectStandAlone.todos.findIndex(todo => todo.id === id);
      const todoIndexInTodosArray = todosArray.findIndex(todo => todo.id === id);
      const updatedTodo = { ...todoFromArray, ...updates };
      // 1. Change todo in todosArray
      todosArray[todoIndexInTodosArray] = updatedTodo;
      localStorageData.setDataIntoLocalStorage('todosArray', todosArray);
      // 2. Change todo is project
      projectStandAlone[todoIndexInProject] = updatedTodo;
      localStorageData.setDataIntoLocalStorage(projectStandAlone.name, projectStandAlone);
      // 3. Change project in projectsArray
      projectsArray[projectIndexInArray] = projectStandAlone;
      localStorageData.setDataIntoLocalStorage('projectsArray', projectsArray);
    },

    delete(id) {
      const projectsArray = localStorageData.getDataFromLocalStorage('projectsArray');
      const todosArray = localStorageData.getDataFromLocalStorage('todosArray');
      const todoInArray = todosArray.find(todo => todo.id === id);
      const projectIndexInArray = projectsArray.findIndex(project => project.id === todoInArray.projectID);
      const projectInArray = projectsArray.find(project => project.id === todoInArray.projectID);
      const projectStandAlone = localStorageData.getDataFromLocalStorage(projectInArray.name);
      // 1. Remove todo in todosArray
      const newTodosArray = todosArray.filter(todo => todo.id !== todoInArray.id);
      localStorageData.setDataIntoLocalStorage('todosArray', newTodosArray);
      // 2. Remove todo in project
      const newProjectsTodos = projectStandAlone.todos.filter(todo => todo.id !== todoInArray.id);
      const newProject = { ...projectStandAlone, todos: newProjectsTodos };
      localStorageData.setDataIntoLocalStorage(projectStandAlone.name, newProject);
      // 3. Update projectsArray
      projectsArray[projectIndexInArray] = newProject;
      localStorageData.setDataIntoLocalStorage('projectsArray', projectsArray);
    },
  })
)();

export default todoController;
