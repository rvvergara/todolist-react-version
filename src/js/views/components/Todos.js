import React from 'react';
import { connect } from 'react-redux';
import TodoItem from './TodoItem';
import AddTodoBtn from './AddTodoBtn';
import TodosForm from './TodosForm';
import todosController from '../../controllers/todosController';
import { getDataFromLocalStorage } from '../../controllers/helpers/common/storage';
import { getTodos, addTodo } from '../actions/todos';
export class Todos extends React.Component {
  state = {
    todos: [],
    project: this.props.selectedProject,
    addTodoMode: false,
  }
  
  componentWillMount(){
    const todos = getDataFromLocalStorage('todosArray');
    this.props.getTodos(todos);
  }

  handleTodoBtn = () => {
    this.setState(() => ({
      addTodoMode: true,
    }));
  }

  handleChange = (key, val) => this.setState({
    [key]: val,
  });

  submitTodo = (todoFromForm) => {
    const newTodo = this.props.addTodo(todoFromForm, this.props.selectedProject.id);
    todosController.create(newTodo.todo);
    this.setState({
      addTodoMode: false,
    })
  }

  deleteTodo = (e) => {
    console.log("todo deleted")
  }

  render() {
    return (
      <div>
        {this.state.project ? <h2>
          Todo list for&nbsp;
          { this.props.selectedProject.name }
        </h2> : <h2>Project Deleted</h2>}
        <table
          className="table table-striped"
        >
          <thead>
            <tr>
              <th>Title</th>
              <th>Description</th>
              <th>Due Date</th>
              <th>Priority</th>
              <th>Note</th>
              <th>Status</th>
              <th>Action</th>
            </tr>
          </thead>
          <tbody>
            {
              this.props.shownTodos.map((todo, i) => (
                <TodoItem
                todo={todo}
                key={todo.id}
                deleteTodo={this.deleteTodo}
                />
                ))
            }
          </tbody>
        </table>
        {
          this.props.shownTodos.length === 0 && <div>No todos yet for {this.props.selectedProject.name}</div>
        }
        {this.props.selectedProject ? <AddTodoBtn
        handleTodoBtn={this.handleTodoBtn}
        addTodoMode={this.state.addTodoMode}
        /> : "No todos for non-existent project"}
        {this.state.addTodoMode && <TodosForm
        addTodoMode={this.state.addTodoMode}
        handleSubmit={this.submitTodo}
        handleChange={this.handleChange}
        />}
      </div>
    )
  }
}

const mapStateToProps = state => ({
  todos: state.todos,
  shownTodos: state.todos.filter(todo => todo.projectID === state.selectedProject.id),
  selectedProject: state.selectedProject,
});

export default connect(mapStateToProps, { getTodos, addTodo })(Todos);