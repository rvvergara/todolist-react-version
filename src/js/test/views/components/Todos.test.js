import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';
import { Todos } from '../../../views/components/Todos';
import projects from '../../fixtures/projects';
import todos from '../../fixtures/todos';
import todosController from '../../../controllers/todosController';
import * as storage from '../../../controllers/helpers/common/storage';

describe('Todos', () => {
  let wrapper;
  let addProjectModeSwitch;
  let addTodo;
  let addTodoModeSwitch;
  let deleteTodo;
  let editProjectModeSwitch;
  let editTodoModeSwitch;
  let getTodos;
  let allTodos;

  beforeEach(() => {
    todosController.create = jest.fn();
    addProjectModeSwitch = jest.fn();
    addTodo = jest.fn(() => ({ todo: {} }));
    addTodoModeSwitch = jest.fn();
    deleteTodo = jest.fn();
    editProjectModeSwitch = jest.fn();
    editTodoModeSwitch = jest.fn();
    getTodos = jest.fn();
    wrapper = shallow(
      <Todos
        addProjectModeSwitch={addProjectModeSwitch}
        addTodoModeSwitch={addTodoModeSwitch}
        addTodo={addTodo}
        deleteTodo={deleteTodo}
        editProjectModeSwitch={editProjectModeSwitch}
        editTodoModeSwitch={editTodoModeSwitch}
        getTodos={getTodos}
        todos={todos}
        shownTodos={[todos[0]]}
        selectedProject={projects[0]}
        addTodoMode={false}
        editTodoMode={false}
        addProjectMode={false}
        editProjectMode={false}
      />,
    );
    storage.getDataFromLocalStorage = jest.fn(() => todos);
    allTodos = storage.getDataFromLocalStorage('someName');
  });

  describe('rendering and mounting', () => {
    test('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('should get data from database/storage and call getTodos prior to mounting', () => {
      expect(storage.getDataFromLocalStorage).toHaveBeenLastCalledWith(expect.any(String));
      expect(getTodos).toHaveBeenLastCalledWith(allTodos);
    });
  });

  describe('should handle interaction with the AddTodoBtn component', () => {
    test('it should call addTodoModeSwitch and render TodosForm and not show AddTodo button', () => {
      wrapper.find('AddTodoBtn').prop('handleTodoBtn')();
      expect(addTodoModeSwitch).toHaveBeenCalled();
      wrapper.setProps({ addTodoMode: true });
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find('TodosForm')).toBeTruthy();
      expect(wrapper.find('AddTodoBtn')).toEqual({});
    });

    test('it should call addProjectModeSwitch if addProjectMode is true initially', () => {
      wrapper.setProps({ addProjectMode: true });
      wrapper.find('AddTodoBtn').prop('handleTodoBtn')();
      expect(addProjectModeSwitch).toHaveBeenCalled();
    });

    test('it should call editProjectModeSwitch if editProject mode is initially true', () => {
      wrapper.setProps({ editProjectMode: true });
      wrapper.find('AddTodoBtn').prop('handleTodoBtn')();
      expect(editProjectModeSwitch).toHaveBeenCalled();
    });
  });

  describe('should handle form submission', () => {
    beforeEach(() => {
      wrapper.setProps({ addTodoMode: true });
      wrapper.find('TodosForm').prop('handleSubmit')({});
    });

    test('it should call addTodo', () => {
      expect(addTodo).toHaveBeenLastCalledWith({}, projects[0].id);
    });

    test('it should call todosController.create method', () => {
      expect(todosController.create).toHaveBeenLastCalledWith({});
    });

    test('it should call addToDoModeSwitch', () => {
      expect(addTodoModeSwitch).toHaveBeenCalled();
    });
  });

  describe('should handle deleteTodo', () => {
    test('it should call deleteTodo with the id of the todo to be deleted', () => {
      wrapper.find('ConnectFunction').prop('deleteTodo')(todos[0].id);
      expect(deleteTodo).toHaveBeenLastCalledWith(todos[0].id);
    });

    test('it should call addTodoModeSwitch if addTodoMode is initially true', () => {
      wrapper.setProps({ addTodoMode: true });
      wrapper.find('ConnectFunction').prop('deleteTodo')(todos[0].id);
      expect(addTodoModeSwitch).toHaveBeenCalled();
    });

    test('should call editTodoModeSwitch if editTodoMode is initially true', () => {
      wrapper.setProps({ editTodoMode: true });
      wrapper.find('ConnectFunction').prop('deleteTodo')(todos[0].id);
      expect(editTodoModeSwitch).toHaveBeenCalled();
    });

    test('should call addProjectModeSwitch if addProjectMode is initially true', () => {
      wrapper.setProps({ addProjectMode: true });
      wrapper.find('ConnectFunction').prop('deleteTodo')(todos[0].id);
      expect(addProjectModeSwitch).toHaveBeenCalled();
    });

    test('should call editProjectModeSwitch if editProjectMode is initially true', () => {
      wrapper.setProps({ editProjectMode: true });
      wrapper.find('ConnectFunction').prop('deleteTodo')(todos[0].id);
      expect(editProjectModeSwitch).toHaveBeenCalled();
    });
  });
});
