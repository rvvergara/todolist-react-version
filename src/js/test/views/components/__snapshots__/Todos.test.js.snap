// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos rendering and mounting should render correctly 1`] = `
<div>
  <h2>
    Todo list for 
    First
  </h2>
  <table
    className="table table-striped"
  >
    <thead>
      <tr>
        <th>
          Title
        </th>
        <th>
          Description
        </th>
        <th>
          Due Date
        </th>
        <th>
          Priority
        </th>
        <th>
          Note
        </th>
        <th>
          Status
        </th>
        <th>
          Action
        </th>
      </tr>
    </thead>
    <tbody>
      <ConnectFunction
        deleteTodo={[Function]}
        key="10"
        todo={
          Object {
            "description": "Sunt incididunt eu exercitation irure magna anim quis laboris incididunt ex occaecat.",
            "done": false,
            "dueDate": 1557035179624,
            "id": 10,
            "notes": "Non anim quis veniam et ea.",
            "projectName": "First",
            "title": "First Task",
          }
        }
      />
    </tbody>
  </table>
  <AddTodoBtn
    addTodoMode={false}
    editTodoMode={false}
    handleTodoBtn={[Function]}
  />
</div>
`;

exports[`Todos should handle interaction with the AddTodoBtn component it should call addTodoModeSwitch and render TodosForm and not show AddTodo button 1`] = `
<div>
  <h2>
    Todo list for 
    First
  </h2>
  <table
    className="table table-striped"
  >
    <thead>
      <tr>
        <th>
          Title
        </th>
        <th>
          Description
        </th>
        <th>
          Due Date
        </th>
        <th>
          Priority
        </th>
        <th>
          Note
        </th>
        <th>
          Status
        </th>
        <th>
          Action
        </th>
      </tr>
    </thead>
    <tbody>
      <ConnectFunction
        deleteTodo={[Function]}
        key="10"
        todo={
          Object {
            "description": "Sunt incididunt eu exercitation irure magna anim quis laboris incididunt ex occaecat.",
            "done": false,
            "dueDate": 1557035179624,
            "id": 10,
            "notes": "Non anim quis veniam et ea.",
            "projectName": "First",
            "title": "First Task",
          }
        }
      />
    </tbody>
  </table>
  <TodosForm
    addTodoMode={true}
    handleSubmit={[Function]}
  />
</div>
`;
